### PR TITLE
Support for bounds-checked indexing of `std::vec`

### PR DESCRIPTION
**Pull request by ranjitjhala**

### CODE REVIEW

File: crates/flux-driver/src/collector.rs [hunk 1]
```diff
@@ -528,6 +541,112 @@ impl<'tcx, 'a> SpecCollector<'tcx, 'a> {
...
+            {
+                return Some(def_id);
+            }
+        }
+        None
+    }
+
+    fn is_good_trait_predicate(trait_predicate: &TraitPredicate) -> bool {
+        let def_id = trait_predicate.trait_ref.def_id;
+        !pretty::def_id_to_string(def_id).contains("Sized") // TODO: use LangItem::Sized?
```
Comment (nilehmann) [inline crates/flux-driver/src/collector.rs:564]: Definitely use `LangItem::Sized` for this

File: crates/flux-driver/src/collector.rs [hunk 2]
```diff
@@ -528,6 +541,112 @@ impl<'tcx, 'a> SpecCollector<'tcx, 'a> {
...
+        Some(impl_id)
+    }
+
+    /// Given as input a fake_method_def_id `fake` where
+    ///     fn fake() where Ty : Trait, {}
+    /// we want to
+    /// 1. extract the [TraitRef] for `<Ty as Trait>` and then
+    /// 2. query [resolve_trait_ref_impl_id] to get the impl_id for the above trait-implementation.
+    /// TODO: sadly the [resolve_trait_ref_impl_id] fails for this? see `extern_spec_impl01.rs`
+    fn _extract_extern_def_id_from_extern_spec_impl_new(
```
Comment (nilehmann) [inline crates/flux-driver/src/collector.rs:623]:
```suggestion
    fn extract_extern_def_id_from_extern_spec_impl_new(
```

File: crates/flux-driver/src/collector.rs [hunk 3]
```diff
@@ -528,6 +541,112 @@ impl<'tcx, 'a> SpecCollector<'tcx, 'a> {
...
+        &mut self,
+        _def_id: LocalDefId,
+        items: &[ImplItemRef],
+    ) -> Option<DefId> {
+        // 1. Find the fake_method's def_id
+        let fake_method_def_id = self._fake_method_of(items)?;
+
+        // 2. Get the fake_method's trait_ref
+        let trait_ref = {
+            // let _generics = self.tcx.generics_of(fake_method_def_id);
```
Comment (nilehmann) [inline crates/flux-driver/src/collector.rs:633]:
```suggestion
```

File: crates/flux-driver/src/collector.rs [hunk 4]
```diff
@@ -528,6 +541,112 @@ impl<'tcx, 'a> SpecCollector<'tcx, 'a> {
...
+            self.tcx
+                .predicates_of(fake_method_def_id)
+                .predicates
+                .iter()
+                .filter_map(|(c, _)| c.as_trait_clause()?.no_bound_vars())
+                .find(|p| Self::is_good_trait_predicate(p))
+                .unwrap()
```
Comment (nilehmann) [inline crates/flux-driver/src/collector.rs:640]:
```suggestion
                ?
```

File: crates/flux-driver/src/collector.rs [hunk 5]
```diff
@@ -609,6 +728,7 @@ enum FluxAttrKind {
     CrateConfig(config::CrateConfig),
     Invariant(surface::Expr),
     Ignore,
+    _FakeImpl,
```
Comment (nilehmann) [inline crates/flux-driver/src/collector.rs:731]:
```suggestion
    FakeImpl,
```

File: crates/flux-driver/src/collector.rs [hunk 6]
```diff
@@ -662,6 +782,10 @@ impl FluxAttrs {
         read_flag!(self, Ignore)
     }
 
+    fn _fake_impl(&mut self) -> bool {
```
Comment (nilehmann) [inline crates/flux-driver/src/collector.rs:785]:
```suggestion
    fn fake_impl(&mut self) -> bool {
```

File: crates/flux-middle/src/global_env.rs [hunk 7]
```diff
@@ -206,8 +206,13 @@ impl<'genv, 'tcx> GlobalEnv<'genv, 'tcx> {
 
     pub fn impl_trait_ref(
         self,
-        impl_id: DefId,
+        impl_id: LocalDefId,
```
Comment (nilehmann) [inline crates/flux-middle/src/global_env.rs:209]: This should also work for external impls so we must take a `DefId`. I'd create a method in `GlobalEnv` like so

```rust
fn extern_id_of(self, def_id: DefId) -> Option<DefId> {
    self.map().extern_id_of(def_id.as_local()?)
}
```

Then the code below changes to

```rust
let impl_id = self.extern_id_of(impl_id).unwrap_or(impl_id);
```

File: crates/flux-refineck/src/constraint_gen.rs [hunk 8]
```diff
@@ -186,6 +186,8 @@ impl<'a, 'genv, 'tcx> ConstrGen<'a, 'genv, 'tcx> {
         // Generate fresh inference variables for refinement arguments
         let refine_args = infcx.instantiate_refine_args(genv, callee_def_id)?;
 
+        // println!("TRACE: check_fn_call (1) {callee_def_id:?} / {generic_args:?} => {fn_sig:?}");
```
Comment (nilehmann) [inline crates/flux-refineck/src/constraint_gen.rs:189]:
```suggestion
```

**Review by nilehmann — APPROVED** [review]

File: crates/flux-driver/src/collector.rs (hunk 5 above)
Comment (ranjitjhala) [reply crates/flux-driver/src/collector.rs:730]: For some odd reason the `fn fake_impl()` method is not visible where I am invoking it, any idea???

File: crates/flux-driver/src/collector.rs (hunk 6 above)
Comment (ranjitjhala) [reply crates/flux-driver/src/collector.rs:784]: see above question.

**Comment by ranjitjhala** [comment]
I can't figure out the `fn fake_impl` mystery (`rustc` refuses to agree its in scope, am probably doing something very silly?)

File: crates/flux-driver/src/collector.rs (hunk 5 above)
Comment (nilehmann) [reply crates/flux-driver/src/collector.rs:726]: Hmm, it may be a false positive. I'll take a look

Comment (nilehmann) [reply crates/flux-driver/src/collector.rs:726]: @ranjitjhala I see what's going on. You were trying to call `fake_impl` on an `&[Attribute]` but that's a method on `FluxAttrs` that you get when you call `parse_flux_attrs`.

**Pull request merged**